Colony Community Tab

rdig opened the pull request:
## Description

This PR adds in a new Community tab to the colony home, which displays the list of users that are "members" of that colony (subscribed to), along with they're respective role.

Note: for the "Admin" role we consider a user with any form of permissions, be they in the root domain, or in any subdomain

**New stuff**

- [x] `Community` dashboard component

**Changes** 

- [x] `ColonyHome` add `Community` tab
- [x] `ColonySubscribe` refetch colony subscribers query
- [x]  `transformers` refactor for reusability
- [x]  `transformers` add `getCommunityRoles`

**Demo**

Community tab, subscribe / unsubscribe:
![demo-community-tab](https://user-images.githubusercontent.com/1193222/73984634-ca107700-4941-11ea-8de6-b2bf347db7c2.gif)

Community tab, empty state:
![demo-community-tab-empty](https://user-images.githubusercontent.com/1193222/73984646-d0065800-4941-11ea-80fb-f5aabcc048f4.gif)

Resolves #2022

ceolson01 commented on src/modules/dashboard/components/Community/Community.tsx:
```diff
@@ -0,0 +1,211 @@
+      const {
+        profile: { walletAddress: userAddress },
+      } = user;
+      let communityRole = 'member';
```
Maybe use an `enum` for these different values?

ceolson01 commented on src/modules/dashboard/components/Community/Community.tsx:
```diff
@@ -0,0 +1,211 @@
+    })
+    .sort(
+      sortObjectsBy({
+        name: 'communityRole',
```
Do we also want to sort these community members by a secondary property beyond the community role? Such as the user's display name, wallet address, etc.?

Also, maybe a question for @JoinColony/product?

collinvine replied:
I had only briefly pondered on that and didn't come up with a great answer.

Ideally, we'd sort on "contribution" or "reputation" but I don't think we'll have that anytime soon.

Any other sorting variable—name, join date—I see pros and cons for. (Name: easy to scan; irrelevant sort | date: oldest at top would be like a kudos, but newest at bottom would push them down)

If there was a forced choice, maybe display name or oldest first (newest last) (that's probably the default, no?). But I'm not sold on any one.

Thoughts?

ceolson01 replied:
I tend to agree with the fact that no other properties are really relevant until reputation comes into the picture

ceolson01 commented on src/modules/transformers.ts:
```diff
@@ -116,35 +116,78 @@ export const getAllUserRoles = (
  */
 export const getLegacyRoles = (
-  domains,
+  domains: Record<string, DomainType>,
```
👍

ceolson01 commented on src/modules/transformers.ts:
```diff
@@ -116,35 +116,78 @@ export const getAllUserRoles = (
-    new Set(),
-  );
+  const founder = getLegacyFounder(rootDomainRoles);
+  const admins = getLegacyAdmins(domains, ROOT_DOMAIN, founder);
```
❤️

ceolson01 commented on src/modules/transformers.ts:
```diff
@@ -116,35 +116,78 @@ export const getAllUserRoles = (
 
+/*
+ * @NOTE This differs from the above transformer as it considers roles in any domain (root + subdomains)
+ * to be an admin role
```
👍

ceolson01 commented on src/modules/dashboard/components/Community/Community.tsx:
```diff
@@ -0,0 +1,211 @@
+  colonyAddress: Address;
+}
+
+const Community = ({ colonyAddress }: Props) => {
```
`displayName` please!

rdig replied:
Granted it's wrong, and an older format introduced by copy-paste, but it is there...

https://github.com/JoinColony/colonyDapp/blob/31e41bdb30de8a4b21461ffe58daaa861a0bbd31/src/modules/dashboard/components/Community/Community.tsx#L209

rdig commented:
@ceolson01 

- Added cache update functions _(although I like this way of saving on server requests, the level of boilerplate code is not something I enjoy, and we need to find a better solution for this)_
- clear the timeout on component unmount
- updated the display name
- added an `enum` to better manage roles within the component

Read for another look.

ceolson01 commented on src/modules/dashboard/components/Community/Community.tsx:
```diff
@@ -0,0 +1,227 @@
+
+const Community = ({ colonyAddress }: Props) => {
+  const [justSubscribed, setJustSubscribed] = useState<boolean>(false);
+  const subscribedMessageTimer = useRef<any>(null);
```
**Optional** This could be typed as a `number`

ceolson01 commented on src/modules/dashboard/components/Community/Community.tsx:
```diff
@@ -0,0 +1,227 @@
+   * function gets called on unmount.
+   * The first one is only called on render.
+   */
+  useEffect(() => () => clearTimeout(subscribedMessageTimer.current), [
```
👍

ceolson01 commented on src/modules/dashboard/components/Community/Community.tsx:
```diff
@@ -0,0 +1,227 @@
+  FOUNDER = 'founder',
+  ADMIN = 'admin',
+  MEMBER = 'member',
+}
```
It seems like we've been tending to prefer pascal case with enums elsewhere. Definitely a nit-pick, and I know we have a couple other instances that are uppercase (likely leftover from js -> ts migration). Totally your call here!

ceolson01 commented on src/data/cacheUpdates.ts:
```diff
@@ -177,6 +183,123 @@ const cacheUpdates = {
+              address: subscribedUserWalletAddress,
+            },
+          });
+          // console.log(subscribedUserProfileFromCache);
```
Leftover comment 🙂

ceolson01 commented on src/data/cacheUpdates.ts:
```diff
@@ -177,6 +183,123 @@ const cacheUpdates = {
+          const { subscribedUsers } = cacheData.colony;
+          /*
+           * The subscribed to colony mutation, only returns the user wallet address,
+           * but we also need the user's profile to update the subscribers array
```
👍
